yarn-project/acir-simulator/src/client/execution_result.ts

```diff
@@ -78,15 +78,8 @@ export interface ExecutionResult {
  * @returns All encrypted logs.
  */
 export function collectEncryptedLogs(execResult: ExecutionResult): FunctionL2Logs[] {
-  const logs: FunctionL2Logs[] = [];
-  // traverse through the stack of nested executions
-  const executionStack = [execResult];
-  while (executionStack.length) {
-    const currentExecution = executionStack.pop()!;
-    executionStack.push(...currentExecution.nestedExecutions);
-    logs.push(currentExecution.encryptedLogs);
-  }
-  return logs;
+  // without the .reverse(), the logs will be in a queue like fashion which is wrong as the kernel processes it like a stack.
+  return [execResult.encryptedLogs, ...execResult.nestedExecutions.reverse().flatMap(collectEncryptedLogs)];
```

Comment:
Javascript's native `reverse` is evil af (as most js methods), since it both reverses in place and returns a reference to the reversed array. So, every time you're calling `collectEncryptedLogs`, you're inadvertently reversing all nested executions.

```
> a = [1,2,3]
[ 1, 2, 3 ]
> a.reverse()
[ 3, 2, 1 ]
> a
[ 3, 2, 1 ]
```

Better make a copy of the array before reversing it, like:

```ts
  return [execResult.encryptedLogs, ...[...execResult.nestedExecutions].reverse().flatMap(collectEncryptedLogs)];
```

Reply:
holy mother of god that makes sense - thanks!

Reply:
> holy mother of god 

That's the typical reaction of every developer when faced with js. Reminded me of [this "oh-no js" moment](https://twitter.com/smpalladino/status/1388260070388031492) I had a while back.